cranelift/codegen/src/machinst/mod.rs

```diff
@@ -354,49 +354,66 @@ impl<T: CompilePhase> CompiledCodeBase<T> {
 
         let relocs = self.buffer.relocs();
         let traps = self.buffer.traps();
-        let labels = self.bb_starts.as_slice();
 
-        let insns = cs.disasm_all(self.buffer.data(), 0x0).map_err(map_caperr)?;
-        for i in insns.iter() {
-            if let Some((n, off)) = labels
-                .iter()
-                .copied()
-                .enumerate()
-                .find(|(_, val)| *val == i.address() as u32)
-            {
-                writeln!(buf, "block{}: ; offset 0x{:x}", n, off)?;
+        // Normalize the block starts to include an initial block of offset 0.
+        let block_starts = if let Some(start) = self.bb_starts.first() {
+            let mut starts = Vec::new();
+            if *start != 0 {
+                starts.push(0);
             }
+            starts.extend_from_slice(&self.bb_starts);
+            starts
+        } else {
+            vec![0]
+        };
 
-            write!(buf, "  ")?;
-
-            let op_str = i.op_str().unwrap_or("");
-            if let Some(s) = i.mnemonic() {
-                write!(buf, "{}", s)?;
-                if !op_str.is_empty() {
-                    write!(buf, " ")?;
+        // Iterate over block regions, to ensure that we always produce block labels
+        let end = self.buffer.data().len();
+        for (n, start) in block_starts.iter().enumerate() {
+            let start = *start as usize;
+            let end = if let Some(next) = block_starts.get(n + 1) {
+                *next as usize
+            } else {
+                end
+            };
```

Comment:
I guess this isn't quite the right types (`usize` vs whatever is in `bb_starts`) but I'd suggest something closer to this:
```suggestion
        block_starts.push(self.buffer.data().len());
        for (&start, &end) in block_starts.iter().zip(block_starts.iter().skip(1)) {
```

Reply:
Great idea!